nginx/nginx.go

```diff
@@ -0,0 +1,233 @@
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	extensionsv1beta1 "k8s.io/client-go/pkg/apis/extensions/v1beta1"
+)
+
+// Nginx contains information for nginx deployment/service pair.
+type Nginx struct {
+	Namespace string
+	// Name of the deployment
+	Name string
+	// Name of the service
+	SVCName string
+	// List of pods that belong to the deployment
+	Pods []*v1.Pod
+
+	kc       kubernetes.Interface
+	labelSel *metav1.LabelSelector
+}
+
+// NewNginx create this nginx deployment/service pair.
+// It waits until all the pods are running.
+func NewNginx(kc kubernetes.Interface, namespace string) (*Nginx, error) {
+	//create random suffix
+	name := fmt.Sprintf("nginx-%s", utilrand.String(5))
+
+	n := &Nginx{
+		Namespace: namespace,
+		Name:      name,
+		SVCName:   name,
+		labelSel: &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"app": name,
+			},
+		},
+		kc: kc,
+	}
+
+	//create nginx deployment
+	nd := n.getDeploymentSpecs()
+	if _, err := kc.ExtensionsV1beta1().Deployments(n.Namespace).Create(nd); err != nil {
+		return nil, fmt.Errorf("error creating %s: %v", n.Name, err)
+	}
+	if err := utils.Retry(10, 5*time.Second, func() error {
+		d, err := kc.ExtensionsV1beta1().Deployments(n.Namespace).Get(n.Name, metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		if d.Status.UpdatedReplicas != d.Status.AvailableReplicas && d.Status.UnavailableReplicas != 0 {
+			return fmt.Errorf("expected: %d ready: %d", d.Status.UpdatedReplicas, d.Status.AvailableReplicas)
+		}
+
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("error creating %s: %v", n.Name, err)
+	}
+
+	//wait for all pods to enter running phase
+	if err := utils.Retry(10, 5*time.Second, func() error {
+		pl, err := kc.CoreV1().Pods(n.Namespace).List(metav1.ListOptions{
+			LabelSelector: metav1.FormatLabelSelector(n.labelSel),
+		})
+		if err != nil {
+			return err
+		}
+
+		if len(pl.Items) == 0 {
+			return fmt.Errorf("there should be non-zero no. of pods")
+		}
+
+		var pods []*v1.Pod
+		for i := range pl.Items {
+			p := &pl.Items[i]
+			if p.Status.Phase != v1.PodRunning {
+				return fmt.Errorf("pod %s not running", p.GetName())
+			}
+
+			pods = append(pods, p)
+		}
+
+		n.Pods = pods
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("error creating %s: %v", n.Name, err)
+	}
+
+	//create nginx service
+	svc := n.getServiceSpecs()
+	if _, err := kc.CoreV1().Services(n.Namespace).Create(svc); err != nil {
+		return nil, fmt.Errorf("error creating service %s: %v", n.SVCName, err)
+	}
+	return n, nil
+}
+
+// Ping pings the nginx service.
+// runs `wget --timeout 5 <nginx-svc-name>`
+// expectedPhase is the phase you expect the ping pods to end up in.
+func (n *Nginx) Ping(expectedPhase v1.PodPhase) error {
```

Comment:
(1) Rather than passing expected phase, you should run commands that explicitly check for certain conditions, and the pods should always succeed. So, if you want wget to fail, make the command such that wget failing == pod succeeding. The pod may have a failure for other system reasons (e.g. node becomes unreachable). Then you can also use a Job with an OnFailure restart policy and a timeout and just watch the api object.

(2) Can you provide an option to run the same test until there are N consecutive successes (where success means "criteria reached"). One succeeded/failed ping isn't a reliable indicator of success IMO.